File: mobile/src/main/java/com/tomaszrykala/midimixerandroidthings/midi/legacy/MidiEvent.java
package com.tomaszrykala.midimixerandroidthings.midi.legacy;

public interface MidiEvent {

    byte DEFAULT_VELOCITY = 64;
    byte DEFAULT_NUMBER = 0;

    enum Type {

        NOTE_OFF(STATUS_NOTE_OFF), NOTE_ON(STATUS_NOTE_ON), CONTROL_CHANGE(STATUS_CONTROL_CHANGE);

        public final byte value;

        Type(byte value) {
            this.value = value;
        }
    }

    byte STATUS_NOTE_OFF = (byte) 0x80;
    byte STATUS_NOTE_ON = (byte) 0x90;
    byte STATUS_POLYPHONIC_AFTERTOUCH = (byte) 0xA0;
    byte STATUS_CONTROL_CHANGE = (byte) 0xB0;
    byte STATUS_PROGRAM_CHANGE = (byte) 0xC0;
    byte STATUS_CHANNEL_PRESSURE = (byte) 0xD0;
    byte STATUS_PITCH_BEND = (byte) 0xE0;
}

File: mobile/src/main/java/com/tomaszrykala/midimixerandroidthings/midi/MidiEvent.kt
package com.tomaszrykala.midimixerandroidthings.midi

import com.tomaszrykala.common.MidiEventType
import kotlin.experimental.and
import kotlin.experimental.or

class MidiEvent constructor(
        private val type: Byte,
        private val channel: Byte,
        vararg private val payload: Byte) {

    val bytes: ByteArray
        get() = ByteArray(payload.size + 1) {
            when (it) {
                0 -> type and STATUS_MASK or (channel and CHANNEL_MASK)
                else -> payload[it - 1]
            }
        }

    companion object {
        private const val STATUS_MASK = 0xF0.toByte()
        private const val CHANNEL_MASK = 0x0F.toByte()

        fun noteOn(channel: Int, note: Int, velocity: Int) =
                MidiEvent(MidiEventType.STATUS_NOTE_ON.byte, channel.toByte(), note.toByte(), velocity.toByte())

        fun noteOff(channel: Int, note: Int, velocity: Int) =
                MidiEvent(MidiEventType.STATUS_NOTE_OFF.byte, channel.toByte(), note.toByte(), velocity.toByte())

        fun controlChange(channel: Int, note: Int, velocity: Int) =
                MidiEvent(MidiEventType.STATUS_CONTROL_CHANGE.byte, channel.toByte(), note.toByte(), velocity.toByte())
    }
}